grr/server/grr_response_server/gui/validate_iap.py

```diff
@@ -0,0 +1,93 @@
+# Copyright 2016 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Sample showing how to validate the Identity-Aware Proxy (IAP) JWT.
+
+This code should be used by applications in Google Compute Engine-based
+environments (such as Google App Engine flexible environment, Google
+Compute Engine, or Google Container Engine) to provide an extra layer
+of assurance that a request was authorized by IAP.
+
+For applications running in the App Engine standard environment, use
+App Engine's Users API instead.
+"""
+# [START iap_validate_jwt]
+import jwt
+import requests
+
+def ValidateIapJwtFromComputeEngine(iap_jwt, cloud_project_number,
+                                         backend_service_id):
+    """Validate an IAP JWT for your (Compute|Container) Engine service.
+
+    Args:
+      iap_jwt: The contents of the X-Goog-IAP-JWT-Assertion header.
+      cloud_project_number: The project *number* for your Google Cloud project.
+          This is returned by 'gcloud projects describe $PROJECT_ID', or
+          in the Project Info card in Cloud Console.
+      backend_service_id: The ID of the backend service used to access the
+          application. See
+          https://cloud.google.com/iap/docs/signed-headers-howto
+          for details on how to get this value.
+
+    Returns:
+      (user_id, user_email, error_str).
+    """
+    expected_audience = '/projects/{}/global/backendServices/{}'.format(
+        cloud_project_number, backend_service_id)
+    return ValidateIapJwt(iap_jwt, expected_audience)
+
+
+def ValidateIapJwt(iap_jwt, expected_audience):
+    try:
+        key_id = jwt.get_unverified_header(iap_jwt).get('kid')
+        if not key_id:
+            return (None, None, '**ERROR: no key ID**')
+        key = GetIapKey(key_id)
+        decoded_jwt = jwt.decode(
+            iap_jwt, key,
+            algorithms=['ES256'],
+            audience=expected_audience)
+        return (decoded_jwt['sub'], decoded_jwt['email'], '')
+    except (jwt.exceptions.InvalidTokenError,
+            requests.exceptions.RequestException) as e:
+        return (None, None, '**ERROR: JWT validation error {}**'.format(e))
+
+
+def GetIapKey(key_id):
+    """Retrieves a public key from the list published by Identity-Aware Proxy,
+    re-fetching the key file if necessary.
+    """
+    key_cache = GetIapKey.key_cache
```

Comment:
Makes sense to me. For my own learning, is there a standard that makes the module-level variable a better pattern than the current implementation?

Reply:
I don't think there's a standard, but rather a convention.
* If you use a module-level variable, things are very explicit. There's a singleton variable \_KEY_CACHE (by convention global variables/constants are named in uppercase), it's defined with a default value and it's intended to be used in the module only (because of the "_" prefix).
* If you use the GetApiKey.key_cache approach, the mechanics under the hood are somewhat more complicated. Effectively, you're assigning an attribute to a function object (you can do this because everything is an object in Python). When I see GetIapKey.key_cache, I first assume that GetIapKey is a class (because of an attribute access) and it's a bit surprising to see that it's actually a function. I need to do a few hops in the source code to understand what's going on. Mostly because assigning custom attributes to a function is something that you rarely (if ever) need to do a Python.

Reply:
👍